### PR TITLE
Pay gas for calldata/initial heap size

### PR DESCRIPTION
**Pull request by hanssv**
https://www.pivotaltracker.com/story/show/161317073

### CODE REVIEW

File: apps/aesophia/test/contracts/state_handling.aes [hunk 1]
```diff
@@ -46,6 +54,9 @@ contract StateHandling =
 
   // remote called
   function look_at(s : state) = ()
+  function return_s1() = "short"
+  function return_s2() = "______longer_string_at_least_32_bytes_long_____"
```
Comment (lucafavatella) [inline apps/aesophia/test/contracts/state_handling.aes]: These seem not needed.

File: apps/aevm/src/aevm_eeevm_state.erl [hunk 2]
```diff
@@ -167,7 +167,9 @@ init_vm(State, Code, Mem, Store) ->
                     {ok, CalldataHeap} = aeso_data:binary_to_heap({tuple, [typerep, Type]}, Calldata,
                                                                   aevm_eeevm_maps:next_id(maps(State2)), HeapSize),
                     {Ptr, State3} = write_heap_value(CalldataHeap, State2),
-                    aevm_eeevm_stack:push(Ptr, State3);
+                    CallDataGasCost = aevm_gas:mem_cost(byte_size(Calldata) div 32),
+                    State4 = set_gas(maps:get(gas, State3) - CallDataGasCost, State3),
```
Comment (lucafavatella) [inline apps/aevm/src/aevm_eeevm_state.erl]: I would expect the gas accounting for the calldata to have already been paid by the caller (either contract call transaction, recursive call, contract create transaction - no contracts create another contract as of now).

**Comment by lucafavatella** [comment]
@hanssv I took the liberty of labelling as WIP - not sure whether you plan changes and needs rebase for conflicts anyway.

**Comment by hanssv** [comment]
Updated and rebased, the gas charging is moved to the caller/sender side as discussed in #core-contract - conflicts with #1682!?

File: docs/release-notes/next/PT-161317073.md [hunk 3]
```diff
@@ -0,0 +1 @@
+* Charge gas also for calldata size/initial heap size when doing contract calls. This affects consensus.
```
Comment (lucafavatella) [inline docs/release-notes/next/PT-161317073.md]: I see this PR changes only for Sophia VM. Could you please detail it?

**Pull request merged**